Rework the listening IP/interface selection code

sledgehammer999 opened the pull request:
Closes #11561

It started as a simple fix for #11561. But as I dug deeper I fixed more and more things. And it became a full blown refactor.

With this PR:
1. As per 1.2.x docs, the default setting (aka all interfaces + all addresses) is to also explicitly set the IPv6_ANY address
2. Now if a user chooses "all addresses" from a specific interface, we tell libtorrent to bind to that specific interface instead of listing to it all the addresses of the interface. This has the benefit that libtorrent will automatically follow any changes to that interface. This is of benefit to VPN users when their connection drops and then they get assigned a new "local" ip.
3. Two new options in the drop down menu for addresses. `All IPv4 Addresses` and `All IPv6 Addresses`.
4. Due to the above, the option "Listen to IPv6 address" is removed. This warrants a WebAPI major version bump (not done in this PR).
5. I discovered a Qt bug concerning the textual representation of IPv6 and scope id. Read the code comments for details. This bug essentially resulted in libtorrent not binding on the set IPv6 address but instead to `0.0.0.0`.


PS: I am not very good with choosing variable/function names.

Chocobo1 commented on src/base/utils/net.cpp:
```diff
@@ -91,6 +92,37 @@ namespace Utils
 
+        QHostAddress canonicalIPv6Addr(const QHostAddress &addr)
+        {
+            // Link-local IPv6 textual address always contains a scope id (or zone index)
```
Does users really binds to link-local IP? Is there some real world use case of that?

sledgehammer999 replied:
Either I use wrong terminology or one of us is confused. User may choose (for whatever reason) to listen to a specific IPv6 local address of their network interface. They may want for traffic to go through only through that specific one.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -1244,65 +1243,84 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
-                        .arg(ip, QString::number(port))
-                    , Log::INFO);
+        const QString firstIP = listeningIPs[0];
+        if (firstIP.isEmpty()) {
```
For what reason it can be empty?

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -1244,65 +1243,84 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
-                        , "e.g: Trying to listen on IP: 192.168.0.1, port: 6881")
-                        .arg(ip, QString::number(port))
-                    , Log::INFO);
+        const QString firstIP = listeningIPs[0];
```
Can't listeningIPs be empty list?

sledgehammer999 replied:
Probably! I'll rework this section again. Truth is that the refactoring went "backwards". I started refactoring this code section first AND then I realized that getListeningIPs() had issues and refactored that too. But I never came back here to rethink yet again the logic.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1244,65 +1243,84 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+        else {
+            auto checkAndAddEndpoint = [&] (const QString &nextIP)
+            {
+                const QHostAddress addr = QHostAddress(nextIP);
```
Is it possible to change `nextIP` type to `QHostAddress` and change `getListeningIPs()` to return `QVector<QHostAddress>`? so we can save this conversion and maybe other places too.

sledgehammer999 replied:
The return can't be `QHostAddress` because it might also return an interface name.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1244,65 +1243,84 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+            endpoints << '[' + QHostAddress(QHostAddress::AnyIPv6).toString() + ']' + portString;
+        }
+        else {
+            auto checkAndAddEndpoint = [&] (const QString &nextIP)
```
maybe rename `nextIP` -> `IP`?

glassez replied:
Agree. `nextIP` is invalid name for this function since the function knows about one ip only. `ip` is better name. But, IIRC, it can contains not only IP... Maybe there can be more appropriate name?

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1244,65 +1243,84 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
-                    , Log::INFO);
-                break;
-            }
+        QStringList endpoints;
```
IMO the below code is long enough that should be moved to its own function.

sledgehammer999 commented:
Guys, I think I addressed all your comments so far. Take a look at the new commit.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1496,6 +1434,83 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+        else {
+            for (const QString &ip : listeningIPs) {
+
+                const QHostAddress addr = QHostAddress(ip);
```
you can use the shorter version: `const QHostAddress addr {ip};`

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1496,6 +1434,83 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
 
+void Session::configureNetInterfaces(lt::settings_pack &settingsPack)
+{
+    if (!m_listenInterfaceConfigured) {
```
This redundant level of indention can be eliminated by returning early on the opposite condition.

glassez replied:
Seems it's a "one-level-up" logic. It should be outside of this method:
```c++
if (!m_listenInterfaceConfigured)
    configureNetInterfaces(settings);
```

sledgehammer999 replied:
@glassez I understand what you're saying, but isn't it safer for `configureNetInterfaces()` to do the self-check? Although in this particular case nothing bad will happen if we re-run this function.

sledgehammer999 replied:
Do you insist on this change?

glassez replied:
No.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1496,6 +1434,83 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+                        outgoingInterfaces << guid;
+                    }
+                    else {
+                        LogMsg(tr("Could not get GUID of: %1").arg(ip) , Log::WARNING);
```
maybe "Could not get GUID of network interface: %1"? (and below too)

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -2416,22 +2431,42 @@ QStringList Session::getListeningIPs() const
+    if (ifaceName.isEmpty() && ifaceAddr.isEmpty())
+        return {}; // Indicates all interfaces + all addresses (aka default)
+
+    const auto checkAndAddIP = [&allIPv4, &allIPv6, &IPs](const QHostAddress &addr, const QHostAddress &match)
```
boolean `allIPv4` & `allIPv6` should be captured by value.

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -2416,22 +2431,42 @@ QStringList Session::getListeningIPs() const
+            || (allIPv6 && (addr.protocol() != QAbstractSocket::IPv6Protocol)))
+            return;
+
+        if (match == addr || allIPv4 || allIPv6)
```
pls add parentheses to the first predicate

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -2416,22 +2431,42 @@ QStringList Session::getListeningIPs() const
+        // At this point ifaceAddr was non-empty
+        // If IPs.isEmpty() it means the configured Address was not found
+        if (IPs.isEmpty()) {
+            LogMsg(tr("qBittorrent didn't find the configured %1 local address to listen on"
```
maybe just "Can not find the configured address \"%1\" to listen on"?

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -2445,40 +2480,24 @@ QStringList Session::getListeningIPs() const
+    // At this point there was an explicit interface and an explicit address set
+    // and the address should have been found
+    if (IPs.isEmpty()) {
+        LogMsg(tr("qBittorrent didn't find the configured %1 local address to listen on"
```
maybe just "Can not find the configured address "%1" to listen on"?

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1496,6 +1434,83 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+        const QStringList listeningIPs = getListeningIPs();
+        const QString portString = ':' + QString::number(port);
+
+        //const QString firstIP = listeningIPs[0];
```
I see no point leaving this comment.

SadPencil commented:
As an IPv6 user, can't wait for this bug being fixed. I hope this fix could be approved fastly.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -1496,6 +1434,83 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+
+        //const QString firstIP = listeningIPs[0];
+        if (listeningIPs.isEmpty()) {
+            endpoints << QHostAddress(QHostAddress::AnyIPv4).toString() + portString;
```
Maybe define some helper like:
```c++
QString makeEndpointString(QHostAddress ip, int port);
```

sledgehammer999 replied:
Nah, I don't think this is worth it. It is only 2-3 places where we make endpoints.

sledgehammer999 replied:
Just to clarify: I don't see a point to this so I am not going to implement it.

glassez replied:
Ok. Don't insist.

glassez commented on src/base/utils/net.cpp:
```diff
@@ -91,6 +92,37 @@ namespace Utils
+            // as scope id. Furthermore, QHostAddress doesn't have any convenient method to
+            // affect this, so we jump through hoops here.
+            if (addr.protocol() != QAbstractSocket::IPv6Protocol)
+                return addr;
```
canonicalIPv6Addr shouldn't return non-IPv6 address!

sledgehammer999 replied:
So return a null QHostAddress?

sledgehammer999 commented:
In this push, I addressed almost all the comments, apart 2 from glassez.

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -295,32 +299,40 @@ void AdvancedSettings::updateInterfaceAddressCombo()
+
         //Try to select the last added one
-        if (ip == currentAddress)
+        const QHostAddress currentAddr(currentAddress); // IPv6 addresses have multiple textual represantations
```
typo: represantations -> representations

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1496,6 +1434,82 @@ void Session::loadLTSettings(lt::settings_pack &settingsPack)
+                // Libtorrent expects GUIDs for the 'listen_interfaces' setting.
+                const QString guid = convertIfaceNameToGuid(ip);
+                if (!guid.isEmpty()) {
+                    endpoints <<( guid + portString);
```
the white space placed wrong: `<<( ` -> `<< (`

glassez commented on src/base/utils/net.cpp:
```diff
@@ -91,6 +92,37 @@ namespace Utils
+            // as scope id. Furthermore, QHostAddress doesn't have any convenient method to
+            // affect this, so we jump through hoops here.
+            if (addr.protocol() != QAbstractSocket::IPv6Protocol)
+                return {};
```
If applying this function to IPv4 address isn't valid "use case" you should prevent it in any way (at least add an appropriate assertion). Otherwise it may allow logically incorrect code.

sledgehammer999 replied:
I am confused. How do I prevent it in any way?
And won't the assertion work only in debug?

glassez replied:
>How do I prevent it in any way?


You can throw RuntimeError...

Or you can modify it to be valid for any IP (and rename it to `canonicalIPAddr`).

glassez replied:
Or even it can return IPv6 representstion of IPv4... unless I am wrong it is possible.

sledgehammer999 replied:
[QHostAddress::toIPv6Address()](https://doc.qt.io/qt-5/qhostaddress.html#toIPv6Address) seems to do the job. So it becomes:
```c++
if (addr.protocol() != QAbstractSocket::IPv6Protocol)
                return {addr.toIPv6Address()};
```

glassez replied:
And also apply all remaining modifications.

glassez replied:
Or it already in "canonical" form?

sledgehammer999 replied:
>Or it already in "canonical" form?

It can't be, because at this point it is IPv4.
If you mean below, there is no way to know beforehand. So we convert unconditionally.

glassez replied:
I would implement anothet variant instead. I mean `canonicalIPAddr` that returns IP of the same type as its argument.
BTW, we can use it unconditionally since it always returns valid result.

glassez commented on src/base/bittorrent/session.cpp:
```diff
@@ -2416,22 +2430,42 @@ QStringList Session::getListeningIPs() const
     const QString ifaceAddr = networkInterfaceAddress();
-    const bool listenIPv6 = isIPv6Enabled();
+    const QHostAddress configuredAddr(ifaceAddr);
+    const bool allIPv4 = (ifaceAddr == QLatin1Char('4')); // Means All IPv4 addresses
```
Maybe use some more IP-specific placeholders (e.g. `0.0.0.0` and `::0`) instead? It appears in configure file and can confuse the readers.

sledgehammer999 replied:
Interesting. I'll take a look into the rest of the code and if it doesn't break something, I'll change it to your suggestion.

sledgehammer999 commented:
I think I addressed everything and I did some furhter refactor. Due to that I implemented it as a separate temporary commit. It will be easier for you to see what I changed.

sledgehammer999 commented on src/base/bittorrent/session.cpp:
```diff
@@ -1445,60 +1445,53 @@ void Session::configureNetworkInterfaces(lt::settings_pack &settingsPack)
 
-    if (listeningIPs.isEmpty()) {
-        endpoints << (QHostAddress(QHostAddress::AnyIPv4).toString() + portString);
-        endpoints << ('[' + QHostAddress(QHostAddress::AnyIPv6).toString() + ']' + portString);
```
This branch isn't needed anymore due to the way "All addresses"/All Ipv4/All IPv6 are returned by getListeningIPs().
So all the code below is shifted one indentation level up.

sledgehammer999 commented:
The missing part is the webui code itself. @Chocobo1 will you be able to plug in the correct js/html changes?

Chocobo1 commented on src/base/bittorrent/session.cpp:
```diff
@@ -2431,17 +2424,25 @@ QStringList Session::getListeningIPs() const
-    const bool allIPv4 = (ifaceAddr == QLatin1Char('4')); // Means All IPv4 addresses
-    const bool allIPv6 = (ifaceAddr == QLatin1Char('6')); // Means All IPv6 addresses
+    const bool allIPv4 = (ifaceAddr == QStringLiteral("0.0.0.0")); // Means All IPv4 addresses
+    const bool allIPv6 = (ifaceAddr == QStringLiteral("::")); // Means All IPv6 addresses
```
AFAIK you are duplicating the same string multiple times, i.e. every QStringLiteral() you constructed is going to be distinct in the binary and thus boating it up. IMO using QLatin1String() is suffice.

Chocobo1 commented on src/base/utils/net.cpp:
```diff
@@ -106,7 +106,7 @@ namespace Utils
             // affect this, so we jump through hoops here.
             if (addr.protocol() != QAbstractSocket::IPv6Protocol)
-                return {};
+                return QHostAddress{addr.toIPv6Address()};
```
is the explicit constructor required? would this work: `return addr.toIPv6Address();`?

sledgehammer999 replied:
>would this work: return addr.toIPv6Address();

No. That constructor is marked as explicit.

And `return {addr.toIPv6Address()};` fails with:
```
error: converting to 'QHostAddress' from initializer list would use explicit constructor
```

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -231,18 +231,18 @@ void AdvancedSettings::saveAdvancedSettings()
     else {
         QHostAddress ifaceAddr(m_comboBoxInterfaceAddress.currentText().trimmed());
-        ifaceAddr.isNull() ? session->setNetworkInterfaceAddress({}) : session->setNetworkInterfaceAddress(ifaceAddr.toString());
+        session->setNetworkInterfaceAddress(ifaceAddr.isNull() ? QString {} : ifaceAddr.toString());
```
would this work: `session->setNetworkInterfaceAddress(ifaceAddr.isNull() ? {} : ifaceAddr.toString());`?

sledgehammer999 replied:
Unfortunately no.
gcc:
```
error: expected primary-expression before '{' token
```
msvc:
```
syntax error: '{'
```

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -649,8 +649,22 @@ void AppController::setPreferencesAction()
-        session->setNetworkInterfaceAddress(ifaceAddress.isNull() ? QString {} : ifaceAddress.toString());
+        const QString value = it.value().toString().trimmed();
+        bool ok = false;
+        uint option = value.toUInt(&ok);
```
IMO you should always use `int` unless doing bits operations.

Chocobo1 replied:
And add `const`.

sledgehammer999 replied:
Damn it. I forgot about const because I was trying something else initially.

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -649,8 +649,22 @@ void AppController::setPreferencesAction()
+        uint option = value.toUInt(&ok);
+
+        if (ok) {
+            if (option == 1)
```
for 3-way or more cases better use a `switch()`.

sledgehammer999 replied:
I wanted to avoid the extra space taken by the `break` statement on each branch.

Chocobo1 commented:
>@Chocobo1 will you be able to plug in the correct js/html changes?

Currently I'm low on free time to write code unfortunately.

Chocobo1 commented on src/webui/api/appcontroller.cpp:
```diff
@@ -651,12 +649,23 @@ void AppController::setPreferencesAction()
-        session->setNetworkInterfaceAddress(ifaceAddress.isNull() ? QString {} : ifaceAddress.toString());
+        const QString value = it.value().toString().trimmed();
+        bool ok = false;
+        uint option = value.toUInt(&ok);
```
On a second thought, the original code of this block should be preserved and should let webUI return the ip address string because now the API doesn't make sense.

sledgehammer999 replied:
And how will the Webui user select "I want all IPv4" addresses for the network interface? (vs all address vs all ipv6 addresses vs specific address)

sledgehammer999 replied:
PS: The values here mirror the ones chosen for GUI combobox. They represent the position of the 3 top items in the combobox list.

Chocobo1 replied:
>And how will the Webui user select "I want all IPv4" addresses for the network interface? 

I guess some modification in js code is required, that is to mimic the new combobox options in GUI.

>The values here mirror the ones chosen for GUI combobox. They represent the position of the 3 top items in the combobox list.

You see the api key here is `current_interface_address`, so unless you document explicitly this magic value in WebAPI document, none of the users of WebAPI could guess what `1` or `2` means and not to mention this semantic change warrants a webapi version bump.

sledgehammer999 replied:
>so unless you document explicitly these magic values in WebAPI document

I can do that. Just point me to the correct place.

>warrants a webapi version bump.

The webapi major version bump will happen anyway because I remove the `listen_on_ipv6_address` key

Chocobo1 replied:
@Piccirello 
Any comment on this discussion? https://github.com/qbittorrent/qBittorrent/pull/11592#discussion_r355530995

glassez replied:
Guys, I really don't understand why you can't accept it as it processed by app core (i.e. `0.0.0.0` and `::`)? Why all these "magic" values of separate type?

sledgehammer999 replied:
In this case `0.0.0.0` and `::` will be treated as magic values from the UI perspective. The UI must not show them as `0.0.0.0` and `::` but as `All IPv4` and `All IPv6` respectively. In this context they have special meaning. And we need another value that will indicate `All addresses` (could be an empty string).

So it is a choice on the set of magic numbers. But in this case, don't we run the risk of web devs ignoring these special cases if we just send `0.0.0.0` and `::` ? (sending 0/1/2) will force them to take a better look (and their users will point that out)

In any case, I am no expert in designing outward facing API nor do I care much about the webapi stuff.
If you find a consensus on how it should be represented via the WebAPI, I'll happily oblige.

Chocobo1 replied:
>Guys, I really don't understand why you can't accept it as it processed by app core (i.e. 0.0.0.0 and ::)? 

That is what I meant in the first post.

>And we need another value that will indicate All addresses (could be an empty string).

An empty string for all address is fine for me or even better IMO: `0.0.0.0,::`

>But in this case, don't we run the risk of web devs ignoring these special cases if we just send 0.0.0.0 and :: ? 

From OS API perspective, these 2 values are not special, they just binds to any IPv4/IPv6 address respectively. And any one who ever worked with network sockets will recognize these values. Or do you have other *special* meaning?

>In any case, I am no expert in designing outward facing API nor do I care much about the webapi stuff.

I'm sorry to have to say this, you're touching webapi here and one mistake here is going to cause inconveniences for webAPI devs/users for some period.

>If you find a consensus on how it should be represented via the WebAPI, I'll happily oblige.

`current_interface_address` can only expect an IPV4/IPv6 address string or an empty string.

As for WebUI, (warning, pseudo code blow)
Here need to add an `if (item === '::') // add "all ipv6" option` to populate the entries:
https://github.com/qbittorrent/qBittorrent/blob/975b44d05ffceb147523cc99f878ce8a5a43920e/src/webui/www/private/views/preferences.html#L1442

And here needs to modified to get the ip address value, probably something like: `if ($('optionalIPAddressToBind').getProperty('value') === 'all ipv6') // set to "::"`
https://github.com/qbittorrent/qBittorrent/blob/975b44d05ffceb147523cc99f878ce8a5a43920e/src/webui/www/private/views/preferences.html#L2095

glassez replied:
>So it is a choice on the set of magic numbers. But in this case, don't we run the risk of web devs ignoring these special cases if we just send `0.0.0.0` and `::`?

In any case it's a bad idea to blindly show any received data. UI level should make postprocessing and use appropriate representation. When magic values are documented there is no difference between numbers and "special" IPs. But as I said before "special" IPs are better for readers (for debugging purposes). So even some "stupid" UI shows it as is its better than see some really meaningless numbers.

P.S. Even if the others vote for the numbers and you decide to use them, you don't need to convert anything here, just compare them in string form.

glassez commented:
>An empty string for all address is fine for me or even better IMO: 0.0.0.0,::

Isn't empty string treated as "any IP" by app core? Then it should be used here as well.

sledgehammer999 commented:
Now the webui receives an empty string to mean all addresses, `0.0.0.0` to mean All IPv4 addresses and `::` to mean All IPv6 addresses. This seems to have the benefit of the least possible code/refactoring.

I also changed how the GUI populates the comboboxes to match the logic of the WebUI comboboxes.

And I probably changed other things here and there. So forgive me but please take another look at the whole thing.

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -295,33 +288,35 @@ void AdvancedSettings::updateInterfaceAddressCombo()
-        //Try to select the last added one
-        if (ip == currentAddress)
-            m_comboBoxInterfaceAddress.setCurrentIndex(m_comboBoxInterfaceAddress.count() - 1);
+        if (addr.protocol() == QAbstractSocket::IPv4Protocol) {
```
Just a suggestion, it can be simplified to:
```c++
const QString str = (addr.protocol() == QAbstractSocket::IPv4Protocol)
    ? addr.toString()
    : Utils::Net::canonicalIPv6Addr(addr).toString();
m_comboBoxInterfaceAddress.addItem(str, str);
```
Assuming there won't be another IP version...

sledgehammer999 replied:
In theory there are other values to the protocol enum: https://doc.qt.io/qt-5/qabstractsocket.html#NetworkLayerProtocol-enum
Practically it is unlikely to have one those other values.
In any case, I am leaning at leaving it they way it is now.

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -227,18 +226,12 @@ void AdvancedSettings::saveAdvancedSettings()
-        ifaceAddr.isNull() ? session->setNetworkInterfaceAddress({}) : session->setNetworkInterfaceAddress(ifaceAddr.toString());
-    }
-    session->setIPv6Enabled(m_checkBoxListenIPv6.isChecked());
+    QHostAddress ifaceAddr(m_comboBoxInterfaceAddress.currentData().toString().trimmed());
```
`const QHostAddress ifaceAddr`

sledgehammer999 commented:
On another note, I just noticed that we do a lot of
```c++
// ifaceAddr is of type QHostAddres
functionThatTakesQString(ifaceAddr.isNull() ? QString {} : ifaceAddr.toString());
```

But I realize that we can collapse it to:
```c++
functionThatTakesQString(ifaceAddr.toString());
```

I'll make the changes later (only code affected by me).

Chocobo1 commented on src/gui/advancedsettings.cpp:
```diff
@@ -227,18 +226,14 @@ void AdvancedSettings::saveAdvancedSettings()
-    session->setIPv6Enabled(m_checkBoxListenIPv6.isChecked());
+    // Construct a QHostAddress to filter malformed strings
+    const QHostAddress ifaceAddr(m_comboBoxInterfaceAddress.currentData().toString().trimmed());
+    session->setNetworkInterfaceAddress(ifaceAddr.toString()); //
```
the trailing `//`, is it leftover?

sledgehammer999 commented:
Closes #11632

sledgehammer999 commented:
@glassez do you approve?

glassez commented:
I would still change canonicalIPv6Addr to suggested canonicalIPAddr form. But you can ignore this opinion.

sledgehammer999 commented:
>I would still change canonicalIPv6Addr to suggested canonicalIPAddr form.

IMO, I don't want to give the impression that the function can be used with IPv4. (Yes, it will work but it shouldn't be used for IPv4).

glassez commented:
>it shouldn't be used for IPv4

You cannot forbid it since function accepts generic host address (function name isn't enough, there can be use case when parameter is bypassed unchecked). You either can break it at runtime (throw an exception) or change it to be valid (have a meaning) for any kind of host address.
If my argument is irrelevant to you, you can merge this PR, I have nothing to add.

sledgehammer999 commented:
No, I meant that it is discouraged through the function name.

Thank you all for the reviewing.

an0n666 commented:
This introduced the tracker not working bug in v 4.2.1
The annnounce goes through and tracker status says "working" but still the torrent gets moved to error section unless you expilcity select "Optional IP address to bind to" to "All IPv4 Addresses"
#11691

FranciscoPombal commented:
@sledgehammer999 @glassez @Chocobo1 
There is new information available about this topic in the libtorrent documentation:
https://github.com/arvidn/libtorrent/pull/4267

Not sure if this warrants any changes/additions to the code introduced in this PR, but I thought I should give the heads-up.
